woocommerce/woocommerce-ios
# Mark order as paid locally after successful client side processing for Interac payments

Part of #6338 


### Description


Right now, it's possible for the customer to be charged multiple times after an Interac payment is captured in the app but fails at subsequent steps (e.g. server-side processing or receipt generation) because the "Collect Payment" CTA is still visible. As a lightweight solution by @toupper in pdfdoF-Ip-p2#comment-1506, we can mark an order as paid locally so that the order isn't eligible for IPP again until the next order fetch. On the next pull-to-refresh or other triggers to fetch the latest order, the order's paid state will be updated based on the server response.

#### Major changes

- Added `OrderAction.markOrderAsPaidLocally` to mark an order as paid by setting `Order.datePaid` to a non-nil date. Even though the completion block isn't currently used, I thought it's nice to have one similar to other actions and it's easier for unit testing
- Updated `CollectOrderPaymentUseCase ` to dispatch `OrderAction.markOrderAsPaidLocally` for Interac payments only
- Added test cases for `CollectOrderPaymentUseCase` and `OrderStore`

### Testing instructions


Prerequisites: the store is in Canada and eligible for IPP ([setup instructions](PdfdoF-D-p2))

We can simulate the server-side processing error manually by updating the capture_terminal_payment endpoint to an invalid path [here](https://href.li/?https://github.com/woocommerce/woocommerce-ios/blob/d91c96b72076fbe58990494a8c53b98c7dfdae88/Networking/Networking/Remote/WCPayRemote.swift#L30) (like adding any character at the end). Any other ways to generate an error after client-side processing like during receipt generation also work.

- Go to the orders tab
- Create an order and tap “Collect Payment” when it becomes available
- Connect to a WisePad 3 reader and pay with an Interac card –> the reader should show the payment as successful
- Remove card –> the app should show an alert about the failure (it doesn’t look very user-friendly right now, improvements will be in a separate subtask for #6338)
- Tap “Back to Order” –> the “Collect Payment” CTA should not be visible anymore. before this PR, the “Collect Payment” CTA remains visible
- Pull down to refresh the order --> the order details UI might change based on the API response, the “Collect Payment” CTA should still not be visible if the site has received the Interac payment from webhook


---
